apps/src/javalab/Javalab.js

```diff
@@ -120,7 +139,117 @@ Javalab.prototype.onContinue = function() {
 };
 
 Javalab.prototype.getCode = function() {
-  return '';
+  // store the file version as the source, as we do in WebLab
+  return this.getCurrentFilesVersionId();
+};
+
+Javalab.prototype.getCurrentFilesVersionId = function() {
+  return project.filesVersionId || this.initialFilesVersionId;
+};
+
+// Called by Javalab when a file has been renamed
+Javalab.prototype.renameProjectFile = function(filename, newFilename) {
+  // no need to rename if the name hasn't changed
+  if (filename === newFilename) {
+    return;
+  }
+  header.showProjectSaving();
+  filesApi.renameFile(
+    filename,
+    newFilename,
+    () => {
+      // at this point project.filesVersionId will be updated and we can save
+      // sources as well.
+      project.save();
+    },
+    () => {
+      console.warn(`Javalab: error file ${filename} not renamed`);
+      project.showSaveError_();
+    }
+  );
 };
 
+Javalab.prototype.onSave = function() {
+  header.showProjectSaving();
+  const filename = getStore().getState().javalab.filename;
+  const editorText = getStore().getState().javalab.editorText;
+  // TODO: enable multi-file
+  filesApi.putFile(
+    filename,
+    editorText,
+    () => {
+      // at this point project.filesVersionId will be updated and we can save
+      // sources as well.
+      project.save();
+    },
+    () => {
+      console.warn(`Javalab: error file ${filename} not renamed`);
+      project.showSaveError_();
+    }
+  );
+};
+
+/**
+ * Load the file entry list and store it as this.fileEntries
+ */
+Javalab.prototype.loadFiles = function(callback) {
+  const onFilesReady = (files, filesVersionId) => {
+    assetListStore.reset(files);
+    const fileEntries = assetListStore.list().map(fileEntry => ({
+      name: fileEntry.filename,
+      url: filesApi.basePath(fileEntry.filename),
+      versionId: fileEntry.versionId
+    }));
+    this.populateFiles(fileEntries, callback);
+    this.initialFilesVersionId = this.initialFilesVersionId || filesVersionId;
+
+    if (filesVersionId !== this.initialFilesVersionId) {
+      // After we've detected the first change to the version, we store this
+      // version id so that subsequent writes will continue to replace the
+      // current version (until the browser page reloads)
+      project.filesVersionId = filesVersionId;
+    }
+  };
+
+  filesApi.getFiles(
+    result => onFilesReady(result.files, result.filesVersionId),
+    xhr => {
+      if (xhr.status === 404) {
+        // No files in this project yet, proceed with an empty file
+        // list and no start version id
+        onFilesReady([], null);
+      } else {
+        console.error('files API failed, status: ' + xhr.status);
```

Comment:
nit: make this error more specific (maybe something like, `FilesApi failed to get files. Status: ${xhr.status}`)